Fix Deployment crashes with NPE in  SequenceFlow

Avinash1423 opened the pull request:
closes #42733

Ensures that SequenceFlowValidator validates  targetRef/sourceRef presence and resolvability to prevent NPE

Avinash1423 committed:
Fix Deployment crashes with NPE in  SequenceFlow

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/camunda/camunda?pullRequest=42798) <br/>All committers have signed the CLA.

korthout commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/SequenceFlowTransformer.java:
```diff
@@ -38,10 +38,26 @@ private void connectWithFlowNodes(
+    if(source==null){
+
+      throw new IllegalStateException("SequenceFlow '%s' has unresolved SourceRef".formatted(element.getId()));
+    }
```
❌ While this seems to make sense, it's generally equivalent to throwing a NullPointerException. Whether it's named `NullPointerException` or `IllegalStateException` doesn't matter much.

❓ Do you see other options to recover from previously deployed processes that accidentally didn't contain a source or target for a sequence flow?

korthout reviewed (changes requested):
Thanks for the great contribution @Avinash1423 🏗️ 

👍 This is a great effort towards resolving this bug!

I noticed a few things that need to be addressed before we can merge this. Please have a look at my comments.

❌ Please add a validation test case that shows that we can now validate against this case. You can take inspiration from [`ValidateProcessTest`](https://github.com/camunda/camunda/blob/main/zeebe/bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ValidateProcessTest.java) to introduce a new `SequenceFlowTest` or something similar.

❌ Please follow [our contributing guide](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md):
- align commits to the [commit guidelines](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#commit-message-guidelines) ([see commitlint error](https://github.com/camunda/camunda/actions/runs/20284595969/job/58294357615?pr=42798))
- format according to [our style guide](https://github.com/camunda/camunda/wiki/Code-Style) ([see spotless:check error](https://github.com/camunda/camunda/actions/runs/20284595969/job/58294357632?pr=42798)), just run `./mvnw spotless:apply` to auto-format

Let me know if you have any questions, or if you need help to drive this to completion.

Avinash1423 committed:
Added test case for SequeceFlowValidator and removed changes in SequenceFlowTransformer

Avinash1423 commented:
@korthout 
 I removed the changes in SequenceFlowTransformer and added a test for SequenceFlowValidator 
(SequenceFlowValidationTest + sequence-flow-missing-target.bpmn). 

I’m running into an issue with the test —Kindly take a look.

korthout commented:
@Avinash1423 Hmm. It's an interesting case, and I don't see how it can be reproduced easily. The issue describes some examples, but those do not appear to cause an NPE.

Rather than immediately adding a validation test, we can go for an engine test. These run an in memory-variation of the engine for speed, and the available client (through `ENGINE`) is not equivalent to the full Java client but give the experience as if you're working with the full product.

In such a test, we can deploy the process model (like below). Add this test to the `zeebe/engine` module. And add your bpmn model to the resources in that module under `processes/`.

<details><summary>Example test</summary>
<p>

```java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */
package io.camunda.zeebe.engine.processing.deployment.model.validation;

import io.camunda.zeebe.engine.util.EngineRule;
import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestWatcher;

public final class SequenceFlowValidationTest {

  public static final @ClassRule EngineRule ENGINE = EngineRule.singlePartition();

  @Rule public final TestWatcher testWatcher = new RecordingExporterTestWatcher();

  @Test
  public void shouldRejectDeploymentWhenNoMessageReferenced() {

    final var rejection =
        ENGINE
            .deployment()
            .withXmlClasspathResource("/processes/sequence-flow-missing-target.bpmn")
            .expectRejection()
            .deploy();

    assert false;
  }
}
```

</p>
</details> 

Now you can try running this. You'll notice that the test uses an assertion failure to ensure the test fails. When it fails, it outputs information from what happened on the log stream to standard out. For example, when the deployment is rejected you'll be able to read the reason there.

Using this, you may be able to explore a few more options and actually trigger the NPE. Make sure to rollback your fix before doing this to actually trigger the NPE. Once you've reproduced the case, we can apply the fix to see that the NPE no longer occurs.

Having said this, I personally do not see how this could be reproduced. Setting `targetRef` to `null`, `" "`, or even pointing it to a dangling message didn't work:
```
    <bpmn:sequenceFlow id="Flow_17y48x4" sourceRef="StartEvent_1" targetRef="Message_0b4h5nl" />
  </bpmn:process>
  <bpmn:message id="Message_0b4h5nl" name="Message_0b4h5nl" />
```

The bug report sadly does not include the process model used to produce this. If you are not able to reproduce it, please report your findings in the issue to help a future reader.

Avinash1423 commented:
@korthout  yes
